File: code/data_structures/hash_table/hashtable.cs
using System;
using System.Collections;  // namespace for hashtable
using System.Linq;
using System.Text;
using System.Threading.Tasks;
/*
 * Part of Cosmos by OpenGenus Foundation'
 *
 * The Hashtable class represents a collection of key-and-value pairs that are organized
 * based on the hash code of the key. It uses the key to access the elements in the collection.
 * */
namespace hashtables
{
    sealed class hTable // a class
    {
        static int key; // key of the hastable
        public bool uniqueness { get; set; } // setting to check if allow uniqueness or not
        Hashtable hashtable = new Hashtable();
        internal void add(string value)
        {
            if(uniqueness)
            {
                if(hashtable.ContainsValue(value))
                {
                    Console.WriteLine("Entry exists");
                }
                else
                {
                    hashtable.Add(++key,value);
                }
            }
            else
            {
                hashtable.Add(++key, value);
            }

        }
        internal void delete(int key)
        {
            if (hashtable.ContainsKey(key))
            {
                Console.WriteLine("Deleted value : " + hashtable[key].ToString());
                hashtable.Remove(key);
            }
            else
            {
                Console.WriteLine("Key not found");
            }
        }
        internal void show()
        {
            foreach(var key in hashtable.Keys)
            {
                Console.WriteLine(key.ToString() + " : " + hashtable[key]);
            }
        }

    }
    class Program
    {
        static void Main(string[] args)
        {
            hTable obj = new hTable();

            // i want uniqueness therefore making property true
            obj.uniqueness = true;

            // adding 3 values
            Console.WriteLine("ADDING EXAMPLE");
            obj.add("terabyte");
            obj.add("sandysingh");
            obj.add("adichat");

            // adding terabyte again to see the exists message
            Console.WriteLine("ADDING 'terabyte' AGAIN");
            obj.add("terabyte");

            // showing all original
            Console.WriteLine("Original Table");
            obj.show();

            // removing value at key = 2
            Console.WriteLine("Deleting value");
            obj.delete(2);

            // showing all new
            Console.WriteLine("New Table");
            obj.show();


            Console.ReadKey(); // in ordr to pause the program
        }
    }
}

File: code/data_structures/hash_table/hash_table.js
/*
 * Part of Cosmos by OpenGenus Foundation
 */
function HashTable(obj) {
  this.length = 0;
  this.items  = {};
  
  for (var p in obj) {
    if (obj.hasOwnProperty(p)) {
      this.items[p] = obj[p];
      this.length++;
    }
  }
  
  this.get = function(key) {
    return this.has(key) ? this.items[key] : undefined;
  };
  
  this.set = function(key, value) {
    var previous;
    
    if (!this.has(key)) {
      this.length++;
    } else {
      previous = this.items[key];
    }
    
    this.items[key] = value;
    return previous;
  };
  
  this.has = function(key) {
    return this.items.hasOwnProperty(key);
  };
  
  this.remove = function(key) {
    var previous;
    
    if (this.has(key)) {
      previous = this.items[key];
      this.length--;
      delete this.items[key];
    }
    
    return previous;
  };
  
  this.keys = function() {
    return Object.keys(this.items);
  };
  
  this.values = function() {
    return Object.values(this.items);
  };
  
  this.each = function(fn) {
    for (var k in this.items) {
      if (this.has(k)) {
        fn(k, this.items[k]);
      }
    }
  };
  
  this.clear = function() {
    this.items = {}
    this.length = 0;
  };
}

